File: backend/app/services/shipment_service.py
import uuid
from datetime import date, datetime

from sqlalchemy import func

from app.extensions import db
from app.models.package import Package
from app.models.shipment import Shipment
from app.models.user import User
from app.services.audit_service import (
    ACTION_SHIPMENT_CREATED,
    ACTION_SHIPMENT_DEPARTED,
    ACTION_SHIPMENT_PACKAGE_ADDED,
    ACTION_SHIPMENT_PACKAGE_REMOVED,
    log_entity_action,
    log_package_action,
    ACTION_PACKAGE_STATUS_UPDATED,
)
from app.services.package_service import add_package_event


def create_shipment(
    *,
    reference: str,
    departure_date: date,
    note: str | None,
    created_by: User | None,
) -> Shipment:
    reference = reference.strip()
    if not reference:
        raise ValueError("Reference is required")

    shipment = Shipment(
        reference=reference,
        departure_date=departure_date,
        note=note.strip() if note else None,
        status="open",
        created_by_id=created_by.id if created_by else None,
    )
    db.session.add(shipment)
    db.session.flush()

    if created_by:
        log_entity_action(
            created_by,
            ACTION_SHIPMENT_CREATED,
            "shipment",
            str(shipment.id),
            f"Created departure {reference}",
            metadata={
                "reference": reference,
                "departure_date": departure_date.isoformat(),
            },
        )

    db.session.commit()
    return shipment


def list_shipments(
    *,
    status: str | None = None,
    limit: int = 50,
    offset: int = 0,
) -> tuple[list[Shipment], int]:
    query = Shipment.query
    if status:
        query = query.filter(Shipment.status == status)
    query = query.order_by(Shipment.departure_date.desc(), Shipment.created_at.desc())
    total = query.count()
    shipments = query.offset(offset).limit(limit).all()
    return shipments, total


def get_shipment(shipment_id: uuid.UUID) -> Shipment | None:
    return db.session.get(Shipment, shipment_id)


def _assert_open(shipment: Shipment) -> None:
    if shipment.status != "open":
        raise ValueError(f"Departure {shipment.reference} has already departed")


def _find_other_open_shipment(package: Package, current: Shipment) -> Shipment | None:
    if not package.shipment_id or package.shipment_id == current.id:
        return None
    other = db.session.get(Shipment, package.shipment_id)
    if other and other.status == "open" and other.id != current.id:
        return other
    return None


def add_package_to_shipment(
    shipment: Shipment,
    package: Package,
    *,
    actor: User | None = None,
) -> Package:
    _assert_open(shipment)

    if package.status != "received":
        raise ValueError(f"{package.tracking_number} is not in received status")

    other = _find_other_open_shipment(package, shipment)
    if other:
        raise ValueError(
            f"{package.tracking_number} is already on departure {other.reference}"
        )

    package.shipment_id = shipment.id
    package.updated_at = datetime.utcnow()

    if actor:
        log_entity_action(
            actor,
            ACTION_SHIPMENT_PACKAGE_ADDED,
            "shipment",
            str(shipment.id),
            f"Added {package.tracking_number} to {shipment.reference}",
            metadata={
                "reference": shipment.reference,
                "package_id": str(package.id),
                "tracking_number": package.tracking_number,
            },
        )

    db.session.commit()
    return package


def add_packages_to_shipment(
    shipment: Shipment,
    package_ids: list,
    *,
    actor: User | None = None,
) -> tuple[list[Package], list[dict]]:
    added: list[Package] = []
    failed: list[dict] = []

    for raw_id in package_ids:
        try:
            pid = uuid.UUID(str(raw_id))
        except (TypeError, ValueError):
            failed.append({"id": str(raw_id), "error": "Invalid package ID"})
            continue

        package = db.session.get(Package, pid)
        if not package:
            failed.append({"id": str(raw_id), "error": "Package not found"})
            continue

        try:
            _assert_open(shipment)
            if package.status != "received":
                raise ValueError(f"{package.tracking_number} is not in received status")
            other = _find_other_open_shipment(package, shipment)
            if other:
                raise ValueError(
                    f"{package.tracking_number} is already on departure {other.reference}"
                )
            package.shipment_id = shipment.id
            package.updated_at = datetime.utcnow()
            added.append(package)
        except ValueError as exc:
            failed.append(
                {
                    "id": str(package.id) if package else str(raw_id),
                    "tracking_number": package.tracking_number if package else None,
                    "error": str(exc),
                }
            )

    if added:
        if actor:
            for package in added:
                log_entity_action(
                    actor,
                    ACTION_SHIPMENT_PACKAGE_ADDED,
                    "shipment",
                    str(shipment.id),
                    f"Added {package.tracking_number} to {shipment.reference}",
                    metadata={
                        "reference": shipment.reference,
                        "package_id": str(package.id),
                        "tracking_number": package.tracking_number,
                    },
                )
        db.session.commit()

    return added, failed


def add_package_by_tracking(
    shipment: Shipment,
    tracking: str,
    *,
    actor: User | None = None,
) -> Package:
    tracking = tracking.strip()
    if not tracking:
        raise ValueError("Tracking number is required")

    package = Package.query.filter(
        func.lower(Package.tracking_number) == tracking.lower()
    ).first()
    if not package and not tracking.upper().startswith("PB-"):
        package = Package.query.filter(
            func.lower(Package.carrier_tracking) == tracking.lower()
        ).first()

    if not package:
        raise ValueError(f"Package {tracking} not found")

    return add_package_to_shipment(shipment, package, actor=actor)


def remove_package_from_shipment(
    shipment: Shipment,
    package: Package,
    *,
    actor: User | None = None,
) -> Package:
    _assert_open(shipment)

    if package.shipment_id != shipment.id:
        raise ValueError(f"{package.tracking_number} is not on this departure")

    package.shipment_id = None
    package.updated_at = datetime.utcnow()

    if actor:
        log_entity_action(
            actor,
            ACTION_SHIPMENT_PACKAGE_REMOVED,
            "shipment",
            str(shipment.id),
            f"Removed {package.tracking_number} from {shipment.reference}",
            metadata={
                "reference": shipment.reference,
                "package_id": str(package.id),
                "tracking_number": package.tracking_number,
            },
        )

    db.session.commit()
    return package


def depart_shipment(
    shipment: Shipment,
    *,
    actor: User | None = None,
    note: str | None = None,
) -> tuple[list[Package], list[dict]]:
    _assert_open(shipment)

    packages = list(shipment.packages)
    if not packages:
        raise ValueError("Add at least one package before departing")

    event_note = note.strip() if note else f"Departure {shipment.reference} — {shipment.departure_date.isoformat()}"
    updated: list[Package] = []
    failed: list[dict] = []

    for package in packages:
        if package.status != "received":
            failed.append(
                {
                    "id": str(package.id),
                    "tracking_number": package.tracking_number,
                    "error": f"Package is {package.status}, expected received",
                }
            )
            continue
        try:
            add_package_event(package, "in_transit", event_note)
            updated.append(package)
        except ValueError as exc:
            failed.append(
                {
                    "id": str(package.id),
                    "tracking_number": package.tracking_number,
                    "error": str(exc),
                }
            )

    if failed:
        db.session.rollback()
        raise ValueError(
            f"Could not depart: {len(failed)} package(s) failed. Fix issues and try again."
        )

    if not updated:
        raise ValueError("No packages could be marked in transit")

    now = datetime.utcnow()
    shipment.status = "departed"
    shipment.departed_at = now
    shipment.updated_at = now

    if actor:
        log_entity_action(
            actor,
            ACTION_SHIPMENT_DEPARTED,
            "shipment",
            str(shipment.id),
            f"Departed {shipment.reference} with {len(updated)} package(s)",
            metadata={
                "reference": shipment.reference,
                "departure_date": shipment.departure_date.isoformat(),
                "package_ids": [str(p.id) for p in updated],
                "failed_count": len(failed),
            },
        )
        for package in updated:
            log_package_action(
                actor,
                ACTION_PACKAGE_STATUS_UPDATED,
                str(package.id),
                f"{package.tracking_number}: received → in transit ({shipment.reference})",
                metadata={
                    "tracking_number": package.tracking_number,
                    "from_status": "received",
                    "to_status": "in_transit",
                    "shipment_id": str(shipment.id),
                    "shipment_reference": shipment.reference,
                },
            )

    db.session.commit()
    return updated, failed


def count_open_shipments() -> int:
    return Shipment.query.filter_by(status="open").count()


def batch_depart_packages(
    package_ids: list,
    *,
    shipment_id: uuid.UUID | None = None,
    reference: str | None = None,
    departure_date: date | None = None,
    note: str | None = None,
    actor: User | None = None,
) -> tuple[Shipment, list[Package]]:
    """Create or use an open departure, assign packages, and mark them in transit."""
    if not package_ids:
        raise ValueError("package_ids must be a non-empty array")
    if len(package_ids) > 500:
        raise ValueError("Cannot depart more than 500 packages at once")

    if shipment_id:
        shipment = get_shipment(shipment_id)
        if not shipment:
            raise ValueError("Departure not found")
        _assert_open(shipment)
    else:
        reference = (reference or "").strip()
        if not reference:
            raise ValueError("Departure reference is required")
        if not departure_date:
            raise ValueError("departure_date is required")
        shipment = Shipment(
            reference=reference,
            departure_date=departure_date,
            note=note.strip() if note else None,
            status="open",
            created_by_id=actor.id if actor else None,
        )
        db.session.add(shipment)
        db.session.flush()
        if actor:
            log_entity_action(
                actor,
                ACTION_SHIPMENT_CREATED,
                "shipment",
                str(shipment.id),
                f"Created departure {reference}",
                metadata={
                    "reference": reference,
                    "departure_date": departure_date.isoformat(),
                    "batch_depart": True,
                },
            )

    packages: list[Package] = []
    for raw_id in package_ids:
        try:
            pid = uuid.UUID(str(raw_id))
        except (TypeError, ValueError):
            db.session.rollback()
            raise ValueError(f"Invalid package ID: {raw_id}")

        package = db.session.get(Package, pid)
        if not package:
            db.session.rollback()
            raise ValueError(f"Package not found: {raw_id}")

        if package.status != "received":
            db.session.rollback()
            raise ValueError(f"{package.tracking_number} is not in received status")

        other = _find_other_open_shipment(package, shipment)
        if other:
            db.session.rollback()
            raise ValueError(
                f"{package.tracking_number} is already on departure {other.reference}"
            )

        package.shipment_id = shipment.id
        package.updated_at = datetime.utcnow()
        packages.append(package)

    if not packages:
        db.session.rollback()
        raise ValueError("No packages to depart")

    event_note = (
        note.strip()
        if note
        else f"Departure {shipment.reference} — {shipment.departure_date.isoformat()}"
    )
    updated: list[Package] = []
    for package in packages:
        add_package_event(package, "in_transit", event_note)
        updated.append(package)

    now = datetime.utcnow()
    shipment.status = "departed"
    shipment.departed_at = now
    shipment.updated_at = now

    if actor:
        for package in packages:
            log_entity_action(
                actor,
                ACTION_SHIPMENT_PACKAGE_ADDED,
                "shipment",
                str(shipment.id),
                f"Added {package.tracking_number} to {shipment.reference}",
                metadata={
                    "reference": shipment.reference,
                    "package_id": str(package.id),
                    "tracking_number": package.tracking_number,
                    "batch_depart": True,
                },
            )
        log_entity_action(
            actor,
            ACTION_SHIPMENT_DEPARTED,
            "shipment",
            str(shipment.id),
            f"Departed {shipment.reference} with {len(updated)} package(s)",
            metadata={
                "reference": shipment.reference,
                "departure_date": shipment.departure_date.isoformat(),
                "package_ids": [str(p.id) for p in updated],
                "batch_depart": True,
            },
        )
        for package in updated:
            log_package_action(
                actor,
                ACTION_PACKAGE_STATUS_UPDATED,
                str(package.id),
                f"{package.tracking_number}: received → in transit ({shipment.reference})",
                metadata={
                    "tracking_number": package.tracking_number,
                    "from_status": "received",
                    "to_status": "in_transit",
                    "shipment_id": str(shipment.id),
                    "shipment_reference": shipment.reference,
                    "batch_depart": True,
                },
            )

    db.session.commit()
    return shipment, updated

File: server/src/services/shipmentService.ts
import { and, count, desc, eq, inArray, sql } from 'drizzle-orm'

import { db } from '../db/index.js'
import { packages, shipments } from '../db/schema/index.js'
import type { packages as packagesTable, shipments as shipmentsTable } from '../db/schema/index.js'
import type { UserRow } from '../db/schema/index.js'
import {
  ACTION_PACKAGE_STATUS_UPDATED,
  ACTION_SHIPMENT_CREATED,
  ACTION_SHIPMENT_DEPARTED,
  ACTION_SHIPMENT_PACKAGE_ADDED,
  ACTION_SHIPMENT_PACKAGE_REMOVED,
  logEntityAction,
  logPackageAction,
} from './auditService.js'
import { addPackageEvent } from './packageService.js'

type ShipmentRow = typeof shipmentsTable.$inferSelect
type PackageRow = typeof packagesTable.$inferSelect

export async function createShipment(options: {
  reference: string
  departureDate: string
  note?: string | null
  createdBy?: UserRow | null
}): Promise<ShipmentRow> {
  const reference = options.reference.trim()
  if (!reference) throw new Error('Reference is required')

  const [shipment] = await db
    .insert(shipments)
    .values({
      id: crypto.randomUUID(),
      reference,
      departureDate: options.departureDate,
      note: options.note?.trim() || null,
      status: 'open',
      createdById: options.createdBy?.id ?? null,
      createdAt: new Date(),
      updatedAt: new Date(),
    })
    .returning()

  if (options.createdBy) {
    await logEntityAction(
      options.createdBy,
      ACTION_SHIPMENT_CREATED,
      'shipment',
      shipment.id,
      `Created departure ${reference}`,
      {
        reference,
        departure_date: options.departureDate,
      },
    )
  }

  return shipment
}

export async function listShipments(options?: {
  status?: string | null
  limit?: number
  offset?: number
}): Promise<[ShipmentRow[], number]> {
  let rows = await db.select().from(shipments).orderBy(desc(shipments.departureDate), desc(shipments.createdAt))
  if (options?.status) rows = rows.filter((s) => s.status === options.status)
  const total = rows.length
  const limit = options?.limit ?? 50
  const offset = options?.offset ?? 0
  return [rows.slice(offset, offset + limit), total]
}

export async function getShipment(shipmentId: string): Promise<ShipmentRow | null> {
  const [shipment] = await db.select().from(shipments).where(eq(shipments.id, shipmentId)).limit(1)
  return shipment ?? null
}

function assertOpen(shipment: ShipmentRow): void {
  if (shipment.status !== 'open') {
    throw new Error(`Departure ${shipment.reference} has already departed`)
  }
}

async function findOtherOpenShipment(pkg: PackageRow, current: ShipmentRow): Promise<ShipmentRow | null> {
  if (!pkg.shipmentId || pkg.shipmentId === current.id) return null
  const other = await getShipment(pkg.shipmentId)
  if (other && other.status === 'open' && other.id !== current.id) return other
  return null
}

export async function addPackageToShipment(
  shipment: ShipmentRow,
  pkg: PackageRow,
  options?: { actor?: UserRow | null },
): Promise<PackageRow> {
  assertOpen(shipment)
  if (pkg.status !== 'received') {
    throw new Error(`${pkg.trackingNumber} is not in received status`)
  }

  const other = await findOtherOpenShipment(pkg, shipment)
  if (other) {
    throw new Error(`${pkg.trackingNumber} is already on departure ${other.reference}`)
  }

  const [updated] = await db
    .update(packages)
    .set({ shipmentId: shipment.id, updatedAt: new Date() })
    .where(eq(packages.id, pkg.id))
    .returning()

  if (options?.actor) {
    await logEntityAction(
      options.actor,
      ACTION_SHIPMENT_PACKAGE_ADDED,
      'shipment',
      shipment.id,
      `Added ${pkg.trackingNumber} to ${shipment.reference}`,
      {
        reference: shipment.reference,
        package_id: pkg.id,
        tracking_number: pkg.trackingNumber,
      },
    )
  }

  return updated
}

export async function addPackagesToShipment(
  shipment: ShipmentRow,
  packageIds: string[],
  options?: { actor?: UserRow | null },
): Promise<[PackageRow[], Array<{ id: string; error: string; tracking_number?: string | null }>]> {
  const added: PackageRow[] = []
  const failed: Array<{ id: string; error: string; tracking_number?: string | null }> = []

  for (const rawId of packageIds) {
    const [pkg] = await db.select().from(packages).where(eq(packages.id, String(rawId))).limit(1)
    if (!pkg) {
      failed.push({ id: String(rawId), error: 'Package not found' })
      continue
    }

    try {
      assertOpen(shipment)
      if (pkg.status !== 'received') {
        throw new Error(`${pkg.trackingNumber} is not in received status`)
      }
      const other = await findOtherOpenShipment(pkg, shipment)
      if (other) {
        throw new Error(`${pkg.trackingNumber} is already on departure ${other.reference}`)
      }
      const [updated] = await db
        .update(packages)
        .set({ shipmentId: shipment.id, updatedAt: new Date() })
        .where(eq(packages.id, pkg.id))
        .returning()
      added.push(updated)
    } catch (err) {
      failed.push({
        id: pkg.id,
        tracking_number: pkg.trackingNumber,
        error: err instanceof Error ? err.message : String(err),
      })
    }
  }

  if (added.length && options?.actor) {
    for (const pkg of added) {
      await logEntityAction(
        options.actor,
        ACTION_SHIPMENT_PACKAGE_ADDED,
        'shipment',
        shipment.id,
        `Added ${pkg.trackingNumber} to ${shipment.reference}`,
        {
          reference: shipment.reference,
          package_id: pkg.id,
          tracking_number: pkg.trackingNumber,
        },
      )
    }
  }

  return [added, failed]
}

export async function addPackageByTracking(
  shipment: ShipmentRow,
  tracking: string,
  options?: { actor?: UserRow | null },
): Promise<PackageRow> {
  const trimmed = tracking.trim()
  if (!trimmed) throw new Error('Tracking number is required')

  let [pkg] = await db
    .select()
    .from(packages)
    .where(sql`lower(${packages.trackingNumber}) = lower(${trimmed})`)
    .limit(1)

  if (!pkg && !trimmed.toUpperCase().startsWith('PB-')) {
    ;[pkg] = await db
      .select()
      .from(packages)
      .where(sql`lower(${packages.carrierTracking}) = lower(${trimmed})`)
      .limit(1)
  }

  if (!pkg) throw new Error(`Package ${tracking} not found`)
  return addPackageToShipment(shipment, pkg, options)
}

export async function removePackageFromShipment(
  shipment: ShipmentRow,
  pkg: PackageRow,
  options?: { actor?: UserRow | null },
): Promise<PackageRow> {
  assertOpen(shipment)
  if (pkg.shipmentId !== shipment.id) {
    throw new Error(`${pkg.trackingNumber} is not on this departure`)
  }

  const [updated] = await db
    .update(packages)
    .set({ shipmentId: null, updatedAt: new Date() })
    .where(eq(packages.id, pkg.id))
    .returning()

  if (options?.actor) {
    await logEntityAction(
      options.actor,
      ACTION_SHIPMENT_PACKAGE_REMOVED,
      'shipment',
      shipment.id,
      `Removed ${pkg.trackingNumber} from ${shipment.reference}`,
      {
        reference: shipment.reference,
        package_id: pkg.id,
        tracking_number: pkg.trackingNumber,
      },
    )
  }

  return updated
}

export async function departShipment(
  shipment: ShipmentRow,
  options?: { actor?: UserRow | null; note?: string | null },
): Promise<[PackageRow[], Array<{ id: string; error: string; tracking_number?: string }>]> {
  assertOpen(shipment)

  const shipmentPackages = await db.select().from(packages).where(eq(packages.shipmentId, shipment.id))
  if (!shipmentPackages.length) throw new Error('Add at least one package before departing')

  const eventNote =
    options?.note?.trim() ||
    `Departure ${shipment.reference} — ${shipment.departureDate}`

  const updated: PackageRow[] = []
  const failed: Array<{ id: string; error: string; tracking_number?: string }> = []

  for (const pkg of shipmentPackages) {
    if (pkg.status !== 'received') {
      failed.push({
        id: pkg.id,
        tracking_number: pkg.trackingNumber,
        error: `Package is ${pkg.status}, expected received`,
      })
      continue
    }
    try {
      await addPackageEvent(pkg, 'in_transit', eventNote)
      const [fresh] = await db.select().from(packages).where(eq(packages.id, pkg.id)).limit(1)
      if (fresh) updated.push(fresh)
    } catch (err) {
      failed.push({
        id: pkg.id,
        tracking_number: pkg.trackingNumber,
        error: err instanceof Error ? err.message : String(err),
      })
    }
  }

  if (failed.length) {
    throw new Error(
      `Could not depart: ${failed.length} package(s) failed. Fix issues and try again.`,
    )
  }

  if (!updated.length) throw new Error('No packages could be marked in transit')

  const now = new Date()
  await db
    .update(shipments)
    .set({ status: 'departed', departedAt: now, updatedAt: now })
    .where(eq(shipments.id, shipment.id))

  if (options?.actor) {
    await logEntityAction(
      options.actor,
      ACTION_SHIPMENT_DEPARTED,
      'shipment',
      shipment.id,
      `Departed ${shipment.reference} with ${updated.length} package(s)`,
      {
        reference: shipment.reference,
        departure_date: shipment.departureDate,
        package_ids: updated.map((p) => p.id),
        failed_count: failed.length,
      },
    )
    for (const pkg of updated) {
      await logPackageAction(
        options.actor,
        ACTION_PACKAGE_STATUS_UPDATED,
        pkg.id,
        `${pkg.trackingNumber}: received → in transit (${shipment.reference})`,
        {
          tracking_number: pkg.trackingNumber,
          from_status: 'received',
          to_status: 'in_transit',
          shipment_id: shipment.id,
          shipment_reference: shipment.reference,
        },
      )
    }
  }

  return [updated, failed]
}

export async function countOpenShipments(): Promise<number> {
  const rows = await db.select({ value: count() }).from(shipments).where(eq(shipments.status, 'open'))
  return Number(rows[0]?.value ?? 0)
}

export async function batchDepartPackages(
  packageIds: string[],
  options?: {
    shipmentId?: string | null
    reference?: string | null
    departureDate?: string | null
    note?: string | null
    actor?: UserRow | null
  },
): Promise<[ShipmentRow, PackageRow[]]> {
  if (!packageIds.length) throw new Error('package_ids must be a non-empty array')
  if (packageIds.length > 500) throw new Error('Cannot depart more than 500 packages at once')

  let shipment: ShipmentRow

  if (options?.shipmentId) {
    const found = await getShipment(options.shipmentId)
    if (!found) throw new Error('Departure not found')
    assertOpen(found)
    shipment = found
  } else {
    const reference = (options?.reference ?? '').trim()
    if (!reference) throw new Error('Departure reference is required')
    if (!options?.departureDate) throw new Error('departure_date is required')
    shipment = await createShipment({
      reference,
      departureDate: options.departureDate,
      note: options.note,
      createdBy: options.actor ?? null,
    })
    if (options?.actor) {
      await logEntityAction(
        options.actor,
        ACTION_SHIPMENT_CREATED,
        'shipment',
        shipment.id,
        `Created departure ${reference}`,
        {
          reference,
          departure_date: options.departureDate,
          batch_depart: true,
        },
      )
    }
  }

  const assigned: PackageRow[] = []
  for (const rawId of packageIds) {
    const [pkg] = await db.select().from(packages).where(eq(packages.id, String(rawId))).limit(1)
    if (!pkg) throw new Error(`Package not found: ${rawId}`)
    if (pkg.status !== 'received') {
      throw new Error(`${pkg.trackingNumber} is not in received status`)
    }
    const other = await findOtherOpenShipment(pkg, shipment)
    if (other) {
      throw new Error(`${pkg.trackingNumber} is already on departure ${other.reference}`)
    }
    const [updated] = await db
      .update(packages)
      .set({ shipmentId: shipment.id, updatedAt: new Date() })
      .where(eq(packages.id, pkg.id))
      .returning()
    assigned.push(updated)
  }

  if (!assigned.length) throw new Error('No packages to depart')

  const eventNote =
    options?.note?.trim() ||
    `Departure ${shipment.reference} — ${shipment.departureDate}`

  const updated: PackageRow[] = []
  for (const pkg of assigned) {
    await addPackageEvent(pkg, 'in_transit', eventNote)
    const [fresh] = await db.select().from(packages).where(eq(packages.id, pkg.id)).limit(1)
    if (fresh) updated.push(fresh)
  }

  const now = new Date()
  await db
    .update(shipments)
    .set({ status: 'departed', departedAt: now, updatedAt: now })
    .where(eq(shipments.id, shipment.id))

  if (options?.actor) {
    for (const pkg of assigned) {
      await logEntityAction(
        options.actor,
        ACTION_SHIPMENT_PACKAGE_ADDED,
        'shipment',
        shipment.id,
        `Added ${pkg.trackingNumber} to ${shipment.reference}`,
        {
          reference: shipment.reference,
          package_id: pkg.id,
          tracking_number: pkg.trackingNumber,
          batch_depart: true,
        },
      )
    }
    await logEntityAction(
      options.actor,
      ACTION_SHIPMENT_DEPARTED,
      'shipment',
      shipment.id,
      `Departed ${shipment.reference} with ${updated.length} package(s)`,
      {
        reference: shipment.reference,
        departure_date: shipment.departureDate,
        package_ids: updated.map((p) => p.id),
        batch_depart: true,
      },
    )
    for (const pkg of updated) {
      await logPackageAction(
        options.actor,
        ACTION_PACKAGE_STATUS_UPDATED,
        pkg.id,
        `${pkg.trackingNumber}: received → in transit (${shipment.reference})`,
        {
          tracking_number: pkg.trackingNumber,
          from_status: 'received',
          to_status: 'in_transit',
          shipment_id: shipment.id,
          shipment_reference: shipment.reference,
          batch_depart: true,
        },
      )
    }
  }

  const [finalShipment] = await db.select().from(shipments).where(eq(shipments.id, shipment.id)).limit(1)
  return [finalShipment!, updated]
}
